source/slang/slang-ir-legalize-array-return-type.cpp

```diff
@@ -89,8 +91,11 @@ void legalizeArrayReturnType(IRModule* module)
             if (func->getResultType()->getOp() == kIROp_ArrayType)
             {
                 makeFuncReturnViaOutParam(builder, func);
+                functionsToProcess.add(func);
             }
         }
     }
+    if(isMetalTarget(targetRequest))
+        specializeAddressSpace(module, functionsToProcess);
```

Comment:
I am not sure if this is a best place to do this.

The intention is to solve a problem for the array return type by applying the specialize-address-space.
But this doesn't look well organized and more like a hack.

Is it possible to solve the array return type in the normal flow where the `specializeAddressSpace()` function will be called?

Reply:
> Is it possible to solve the array return type in the normal flow where the specializeAddressSpace() function will be called?

From what I understand the current additional logic is the only non-hack solution without disjointing metal and non metal logic.
`moveGlobalVarInitializationToEntryPoints()` and  `introduceExplicitGlobalContext()` relies upon legalization of `legalizeIRForMetal(irModule, sink);`. `legalizeArrayReturnType()` relies upon legalization of `moveGlobalVarInitializationToEntryPoints()` and `introduceExplicitGlobalContext()`.

> The intention is to solve a problem for the array return type by applying the specialize-address-space.
But this doesn't look well organized and more like a hack.

The problem being solved is that every function which is automatically generated (or modifies parameter list of a function) requires to have their address space re-legalized. This is required since else emitting code may lack the address space information needed to emit Metal code.

`specializeAddressSpace(module);` can be called instead of `specializeAddressSpace(module, functionsToProcess);` as a legalization pass, but this means we will run an IR pass over the entire program instead of just the functions which need re-legalization.

Reply:
The problem might be on the fact that `IRFunc* specializeFunc(const FuncSpecializationKey& key)` handles only the function arguments not the return type.
```
for (auto param : specializedFunc->getParams())
```
But I don't think we can actually handle the return type like how the parameters are handled.
So I guess it is fine as how you did.

Reply:
> The problem might be on the fact that IRFunc* specializeFunc(const FuncSpecializationKey& key) handles only the function arguments not the return type.

`kIROp_Return` and `kIROp_Call` both manage the return type. This works properly with pointers (and pointer-like types: #4465)